### PR TITLE
Fix ready unready

### PR DESCRIPTION
**Pull request by cat-in-vacuum**

### CODE REVIEW

File: dialer.go [hunk 1]
```diff
@@ -153,16 +165,81 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+			if cap(receiver) == 0 {
+				panic(errChanIsUnbuffered)
+			}
+		}
+
+		c.notifyReady = append(c.notifyReady, receiver)
+	}
+}
+
+func WithUnreadyChan(receiver chan error) Option {
```
Comment (makasim) [inline dialer.go:183]: rename receiver -> unreadyCh

File: dialer.go [hunk 2]
```diff
@@ -153,16 +165,81 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+	f := func(c *Dialer) {
+
```
Comment (makasim) [inline dialer.go:185]: in `WithUnreadyChan` you should append channel as is, without any additional checks. Later, [after you've applied options](https://github.com/makasim/amqpextra/blob/50be62389638a2a4dbc41c5f9ecf907890ea1c04/dialer.go#L98) you can check for nil and cap and return an error.

Comment (makasim) [reply dialer.go:185]: same applies to readyCh

File: dialer.go (hunk 1 above)
Comment (makasim) [reply dialer.go:183]: same applies to readyCh

File: dialer.go [hunk 3]
```diff
@@ -12,6 +13,10 @@ import (
 	"github.com/streadway/amqp"
 )
 
+const (
+	errChanIsUnbuffered = "receiver chan is unbuffered"
```
Comment (makasim) [inline dialer.go:17]: create separate errors:

```go
const (
  errReadyChanUnbuffered = fmt.Errorf("ready chan is unbuffered")
  errUnreadyChanUnbuffered = fmt.Errorf("unready chan is unbuffered")
)
```

File: dialer.go [hunk 4]
```diff
@@ -153,16 +165,81 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+	if cap(receiver) == 0 {
+		panic(errChanIsUnbuffered)
+	}
+
+	select {
+	case <- c.NotifyClosed():
+		return receiver
+	default:
+		c.mu.Lock()
+		if receiver == nil {
```
Comment (makasim) [inline dialer.go:213]: do not check for nil

File: dialer.go [hunk 5]
```diff
@@ -153,16 +165,81 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+func (c *Dialer) NotifyReady(receiver chan struct{}) <-chan struct{} {
+	if cap(receiver) == 0 {
+		panic(errChanIsUnbuffered)
+	}
+
+	select {
+	case <- c.NotifyClosed():
+		return receiver
+	default:
+		c.mu.Lock()
```
Comment (makasim) [inline dialer.go:212]: do `defer c.mu.Unlock()` as the next line after lock

File: dialer.go [hunk 6]
```diff
@@ -55,6 +60,11 @@ type Dialer struct {
 
 	connCh    chan *Connection
 	readyCh   chan struct{}
+
+	mu *sync.Mutex
+	notifyReady []chan struct{}
```
Comment (makasim) [inline dialer.go:65]: rename readyChs, unreadyChs, remove unreadyCh, readyCh

File: dialer.go [hunk 7]
```diff
@@ -244,7 +321,9 @@ func (c *Dialer) connectState() {
 		for {
 			select {
 			case c.unreadyCh <- connErr:
-				continue
+				for _, receiver := range c.notifyUnready {
```
Comment (makasim) [inline dialer.go:324]: create a method `func notifyUnready(err error)` that would

* lock
* loop channel,
* do non-blocking channel write

**Review by makasim — CHANGES_REQUESTED** [review]

File: dialer.go (hunk 3 above)
Comment (cat-in-vacuum) [reply dialer.go:17]: fmt.Errorf() returns an error then cannot be used as const. I suggest use something like errMsgDescription for this approach

Comment (makasim) [reply dialer.go:17]: errors.New()

File: dialer.go (hunk 1 above)
Comment (cat-in-vacuum) [reply dialer.go:183]: +

File: dialer.go (hunk 2 above)
Comment (cat-in-vacuum) [reply dialer.go:185]: +

File: dialer.go (hunk 1 above)
Comment (cat-in-vacuum) [reply dialer.go]: renamed

File: dialer.go (hunk 2 above)
Comment (cat-in-vacuum) [reply dialer.go]: checks removed

File: dialer.go (hunk 3 above)
Comment (cat-in-vacuum) [reply dialer.go]: error msgs separated

File: dialer.go (hunk 4 above)
Comment (cat-in-vacuum) [reply dialer.go]: fixed

File: dialer.go (hunk 5 above)
Comment (cat-in-vacuum) [reply dialer.go:222]: fixed

File: dialer.go (hunk 6 above)
Comment (cat-in-vacuum) [reply dialer.go]: chans renamed

File: dialer.go (hunk 7 above)
Comment (cat-in-vacuum) [reply dialer.go]: the method implemented for both ready that unready

File: dialer.go [hunk 8]
```diff
@@ -12,6 +13,12 @@ import (
 	"github.com/streadway/amqp"
 )
 
+
+const (
+	errMsgReadyChanUnbuffered   = "ready chan is unbuffered"
+	errMsgUnreadyChanUnbuffered = "unready chan is unbuffered"
```
Comment (makasim) [inline dialer.go:19]: if something starts with `err` I would expect an error, not a string. Could you create an error using `errors.New("")`?

File: dialer.go [hunk 9]
```diff
@@ -55,6 +62,11 @@ type Dialer struct {
 
 	connCh    chan *Connection
 	readyCh   chan struct{}
```
Comment (makasim) [inline dialer.go:64]: remove

File: dialer.go [hunk 10]
```diff
@@ -55,6 +62,11 @@ type Dialer struct {
...
+
+	mu         *sync.Mutex
+	readyChs   []chan struct{}
+	unreadyChs []chan error
+
 	unreadyCh chan error
```
Comment (makasim) [inline dialer.go:70]: remove

File: dialer.go [hunk 11]
```diff
@@ -97,6 +111,27 @@ func NewDialer(opts ...Option) (*Dialer, error) {
 		opt(c)
 	}
 
+	// if the chan of receivers == nil the checking will be skipped
+	for _, unreadyReceiver := range c.unreadyChs {
+		if unreadyReceiver == nil {
```
Comment (makasim) [inline dialer.go:116]: return err if ch is nil, do not create channel because nobody would able to read from it

File: dialer.go [hunk 12]
```diff
@@ -153,16 +188,59 @@ func WithConnectionProperties(props amqp.Table) Option {
 	}
 }
 
+func WithReadyChan(receiver chan struct{}) Option {
+	f := func(c *Dialer) {
+		c.mu.Lock()
+		defer c.mu.Unlock()
```
Comment (makasim) [inline dialer.go:194]: there is no need for lock here.

File: dialer.go [hunk 13]
```diff
@@ -153,16 +188,59 @@ func WithConnectionProperties(props amqp.Table) Option {
 	}
 }
 
+func WithReadyChan(receiver chan struct{}) Option {
```
Comment (makasim) [inline dialer.go:191]: rename to readyCh

File: dialer.go [hunk 14]
```diff
@@ -153,16 +188,59 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+	f := func(c *Dialer) {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.readyChs = append(c.readyChs, receiver)
+	}
+	return f
+}
+
+func WithUnreadyChan(chUnready chan error) Option {
```
Comment (makasim) [inline dialer.go:200]: rename to unreadyCh

File: dialer.go [hunk 15]
```diff
@@ -153,16 +188,59 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+	return f
+}
+
 func (c *Dialer) ConnectionCh() <-chan *Connection {
 	return c.connCh
 }
 
-func (c *Dialer) NotifyReady() <-chan struct{} {
-	return c.readyCh
+func (c *Dialer) NotifyReady(chReady chan struct{}) <-chan struct{} {
```
Comment (makasim) [inline dialer.go:213]: rename to readyCh

File: dialer.go [hunk 16]
```diff
@@ -153,16 +188,59 @@ func WithConnectionProperties(props amqp.Table) Option {
...
+	select {
+	case <- c.NotifyClosed():
+		close(chUnready)
+	default:
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.unreadyChs = append(c.unreadyChs, chUnready)
+	}
+
+	return chUnready
```
Comment (makasim) [inline dialer.go:243]: move return to `case <- c.NotifyClosed():`

File: dialer.go [hunk 17]
```diff
@@ -301,6 +380,26 @@ func (c *Dialer) connectedState(amqpConn AMQPConnection) error {
...
 
+func(c *Dialer) notifyUnready(err error) {
+	go func(){
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		for _, unreadyReceiver := range c.unreadyChs {
+			unreadyReceiver <- err
+		}
+	}()
+}
```
Comment (makasim) [inline dialer.go:391]:
```go
func(c *Dialer) notifyUnready(err error) {
		c.mu.Lock()
		defer c.mu.Unlock()
		for _, ch := range c.unreadyChs {
                        select {
                        case ch <- err:
                        default:
                        }
		}
}
```

**Review by makasim — CHANGES_REQUESTED** [review]

File: dialer.go (hunk 8 above)
Comment (cat-in-vacuum) [reply dialer.go]: const removed

File: dialer.go (hunk 9 above)
Comment (cat-in-vacuum) [reply dialer.go]: removed

File: dialer.go (hunk 10 above)
Comment (cat-in-vacuum) [reply dialer.go]: removed

File: dialer.go (hunk 11 above)
Comment (cat-in-vacuum) [reply dialer.go]: changed to return error if channel == nil

File: dialer.go (hunk 12 above)
Comment (cat-in-vacuum) [reply dialer.go]: removed

File: dialer.go (hunk 13 above)
Comment (cat-in-vacuum) [reply dialer.go]: renamed

File: dialer.go (hunk 14 above)
Comment (cat-in-vacuum) [reply dialer.go]: renamed

File: dialer.go (hunk 15 above)
Comment (cat-in-vacuum) [reply dialer.go]: renamed

File: dialer.go (hunk 16 above)
Comment (cat-in-vacuum) [reply dialer.go]: Left as it is

File: dialer.go (hunk 17 above)
Comment (cat-in-vacuum) [reply dialer.go:373]: fixed

File: dialer.go [hunk 18]
```diff
@@ -209,24 +270,26 @@ func (c *Dialer) Publisher(opts ...publisher.Option) (*publisher.Publisher, erro
 func (c *Dialer) connectState() {
 	defer close(c.connCh)
 	defer close(c.closedCh)
-	defer close(c.unreadyCh)
```
Comment (makasim) [inline dialer.go:212]: we should close all unreadyChs

File: dialer.go [hunk 19]
```diff
@@ -209,24 +270,26 @@ func (c *Dialer) Publisher(opts ...publisher.Option) (*publisher.Publisher, erro
...
 	defer c.logger.Printf("[DEBUG] connection closed")
 
 	i := 0
 	l := len(c.amqpUrls)
 
 	c.logger.Printf("[DEBUG] connection unready")
+
 	var connErr error = amqp.ErrClosed
+
+	c.notifyUnready(connErr)
```
Comment (makasim) [inline dialer.go:283]: there is no need for connErr anymore.  Please do 
```suggestion
	c.notifyUnready(amqp.ErrClosed)
```

File: dialer_test.go [hunk 20]
```diff
@@ -195,16 +207,18 @@ func TestConnectState(main *testing.T) {
 
 		dialer, err := amqpextra.NewDialer(
 			amqpextra.WithURL("amqp://rabbitmq.host"),
-			amqpextra.WithAMQPDial(amqpDialStub(time.Millisecond*150, conn)),
+			amqpextra.WithAMQPDial(amqpDialStub(conn)),
```
Comment (makasim) [inline dialer_test.go:210]: Revert `time.Millisecond*150`.  The behavior is different without it

File: dialer_test.go [hunk 21]
```diff
@@ -259,21 +276,27 @@ func TestConnectState(main *testing.T) {
...
 			amqpextra.WithURL("amqp://rabbitmq.host"),
-			amqpextra.WithAMQPDial(amqpDialStub(fmt.Errorf("the error"))),
+			amqpextra.WithUnreadyCh(unreadyCh),
 			amqpextra.WithRetryPeriod(time.Millisecond*150),
+			amqpextra.WithAMQPDial(amqpDialStub(fmt.Errorf("the error"))),
 			amqpextra.WithLogger(l),
 		)
 		require.NoError(t, err)
 
 		time.Sleep(time.Millisecond * 100)
```
Comment (makasim) [inline dialer_test.go:290]: remove time.Sleep

File: dialer_test.go [hunk 22]
```diff
@@ -259,21 +276,27 @@ func TestConnectState(main *testing.T) {
...
-		assertUnready(t, dialer, "the error")
 
+		assertUnready(t, unreadyCh, amqp.ErrClosed.Error())
 		dialer.Close()
 		assertClosed(t, dialer)
 
+		time.Sleep(time.Millisecond * 100)
+
+		assertUnready(t, unreadyCh, "the error")
```
Comment (makasim) [inline dialer_test.go:298]: do lines 296-298 after `assertUnready(amqp.ErrClose)` and before `dialer.Close()`

File: dialer_test.go [hunk 23]
```diff
@@ -286,25 +309,31 @@ func TestConnectState(main *testing.T) {
 
 		ctrl := gomock.NewController(t)
 		defer ctrl.Finish()
-
+		// TODO
```
Comment (makasim) [inline dialer_test.go:312]: is it still relevant ?

File: dialer_test.go [hunk 24]
```diff
@@ -286,25 +309,31 @@ func TestConnectState(main *testing.T) {
...
 			amqpextra.WithURL("amqp://rabbitmq.host"),
 			amqpextra.WithAMQPDial(amqpDialStub(fmt.Errorf("the error"))),
 			amqpextra.WithRetryPeriod(time.Millisecond*150),
+			amqpextra.WithUnreadyCh(unreadyCh),
 			amqpextra.WithLogger(l),
 			amqpextra.WithContext(ctx),
 		)
 		require.NoError(t, err)
 
 		time.Sleep(time.Millisecond * 100)
```
Comment (makasim) [inline dialer_test.go:328]: remove timer

File: dialer_test.go [hunk 25]
```diff
@@ -286,25 +309,31 @@ func TestConnectState(main *testing.T) {
...
-		assertUnready(t, dialer, "the error")
+		assertUnready(t, unreadyCh, amqp.ErrClosed.Error())
 
 		cancelFunc()
+
+		time.Sleep(time.Millisecond * 100)
+
+		assertUnready(t, unreadyCh, "the error")
```
Comment (makasim) [inline dialer_test.go:335]: do lines 333-335 after assertUnready(amqp.ErrClose) and before cancelFunc()

File: dialer_test.go [hunk 26]
```diff
@@ -644,14 +694,41 @@ func TestConnectedState(main *testing.T) {
 [DEBUG] connection closed
 `, l.Logs())
 	})
+
+	main.Run("UrlsIterEqualUrlsOrder", func(t *testing.T) {
```
Comment (makasim) [inline dialer_test.go:698]: UrlOrderInDial

File: dialer_test.go [hunk 27]
```diff
@@ -644,14 +694,41 @@ func TestConnectedState(main *testing.T) {
...
+		defer goleak.VerifyNone(t)
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		urlsPool := []string{
+			"the.first.url",
+			"the.second.url",
+			"the.last.url",
+		}
+		wantedIndex := 0
```
Comment (makasim) [inline dialer_test.go:708]: wantedIndex -> index

File: dialer_test.go [hunk 28]
```diff
@@ -644,14 +694,41 @@ func TestConnectedState(main *testing.T) {
...
+		dialSub := func(url string, config amqp.Config) (amqpextra.AMQPConnection, error) {
+			assert.Equal(t, url, urlsPool[wantedIndex])
```
Comment (makasim) [inline dialer_test.go:710]: assert.Equal -> require.Equal

File: dialer_test.go [hunk 29]
```diff
@@ -644,14 +694,41 @@ func TestConnectedState(main *testing.T) {
...
+			wantedIndex++
+
+			return nil, errors.New("the error")
```
Comment (makasim) [inline dialer_test.go:713]:
```suggestion
			return nil, errors.New("let's dial again")
```

File: dialer_test.go [hunk 30]
```diff
@@ -644,14 +694,41 @@ func TestConnectedState(main *testing.T) {
...
+		}
+
+		dialer, err := amqpextra.NewDialer(
+			amqpextra.WithURL(urlsPool[0], urlsPool...),
+			amqpextra.WithAMQPDial(dialSub),
+		)
+
+		require.NoError(t, err)
+
+		dialer.Close()
```
Comment (makasim) [inline dialer_test.go:723]: you have to wait here for sometime, cuz now it exists after first dial. 

So, set some retryPeriod and wait for 2.5 of its time.

**Review by makasim — CHANGES_REQUESTED** [review]

File: dialer.go (hunk 18 above)
Comment (cat-in-vacuum) [reply dialer.go:212]: fixed

File: dialer.go (hunk 19 above)
Comment (cat-in-vacuum) [reply dialer.go]: +

File: dialer_test.go (hunk 20 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go (hunk 21 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go (hunk 22 above)
Comment (cat-in-vacuum) [reply dialer_test.go:305]: +

File: dialer_test.go (hunk 23 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: not actual. Removed

File: dialer_test.go (hunk 24 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go (hunk 25 above)
Comment (cat-in-vacuum) [reply dialer_test.go:341]: +

File: dialer_test.go (hunk 26 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go (hunk 27 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go (hunk 28 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go (hunk 29 above)
Comment (cat-in-vacuum) [reply dialer_test.go:720]: +

File: dialer_test.go (hunk 30 above)
Comment (cat-in-vacuum) [reply dialer_test.go]: +

File: dialer_test.go [hunk 31]
```diff
@@ -644,14 +702,42 @@ func TestConnectedState(main *testing.T) {
...
+		}
+		index := 0
+		dialSub := func(url string, config amqp.Config) (amqpextra.AMQPConnection, error) {
+			require.Equal(t, url, urlsPool[index])
+			index++
+			return nil, errors.New("the error")
+		}
+
+		d, err := amqpextra.NewDialer(
+			amqpextra.WithRetryPeriod(time.Millisecond*100),
```
Comment (makasim) [inline dialer_test.go:724]: set to 20 milliseconds

File: dialer_test.go [hunk 32]
```diff
@@ -644,14 +702,42 @@ func TestConnectedState(main *testing.T) {
...
+			amqpextra.WithURL(urlsPool...),
+			amqpextra.WithAMQPDial(dialSub),
+		)
+		require.NoError(t, err)
+
+		time.Sleep(time.Millisecond * 300)
```
Comment (makasim) [inline dialer_test.go:730]: sleep for 70 milliseconds

File: dialer_test.go [hunk 33]
```diff
@@ -644,14 +702,42 @@ func TestConnectedState(main *testing.T) {
 [DEBUG] connection closed
 `, l.Logs())
 	})
+
+	main.Run("UrlOrderInDial", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		urlsPool := []string{
```
Comment (makasim) [inline dialer_test.go:711]: urlsPool -> expectedUrls

File: dialer_test.go [hunk 34]
```diff
@@ -750,6 +835,10 @@ func any() gomock.Matcher {
 func amqpDialStub(conns ...interface{}) func(url string, config amqp.Config) (amqpextra.AMQPConnection, error) {
 	index := 0
 	return func(url string, config amqp.Config) (amqpextra.AMQPConnection, error) {
+		log.Printf("dial stub called %d time\n", index+1)
```
Comment (makasim) [inline dialer_test.go:838]: remove

File: dialer_test.go [hunk 35]
```diff
@@ -750,6 +835,10 @@ func any() gomock.Matcher {
...
+		if index == len(conns) {
+			panic(fmt.Sprintf("dial stub called more times(%d) than len(conns) - %d", index+1, len(conns)))
```
Comment (makasim) [inline dialer_test.go:840]: dial stub must be called not more than %d times.

**Review by makasim — CHANGES_REQUESTED** [review]

**Review by makasim — APPROVED** [review]

**Pull request merged**